### ISSUE TITLE
Wiki page explaining commonly encountered Server Errors and how to prevent them

### ISSUE THREAD

**Issue by kevintab95** [oppia/oppia-web-developer-docs#60]
This document contains an overview of the server errors encountered in this project and lists some ways to prevent them in the future. It would be good to have this on the wiki page for better visibility to the community and so we can learn from it (and add to it if we see any new patterns / fixes) 

https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.4nsx9mmay1nn

**Comment by Patel-Muhammad** [comment]
Hey @kevintab95 can I take on this issue?

**Comment by U8NWXD** [comment]
@Patel-Muhammad As requested in our [getting started wiki page](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#choosing-a-good-first-issue), please make sure your comment does the following:

> * describes in more detail how you'd tackle the issue (e.g. explain which file(s) you would modify and what changes you would make). If your explanation makes sense, we'll assign the issue to you.

**Comment by Patel-Muhammad** [comment]
## @U8NWXD Sorry, my apologies. 
### I will do the following:
**1. Create a Server-errors-and-preventions.md file.**

**2. I will write about server errors from the [link](https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.4nsx9mmay1nn)**

**3. I will write the following in the .md file:**

This document presents an overview of different categories of server errors, based on errors that have been fixed in the past, and proposes ways to prevent such errors from occurring in the future. The categories include: Data-related issues, Scalability issues, Corner cases, Supporting old browsers, Logic errors, Race conditions/issues on slow networks, and Errors that didn't need to be logged. For each category, an example of an error is provided and preventive actions are suggested.

Server Errors are a common issue faced by developers and can be caused by a variety of factors. In this document, we will present the main categories of server errors based on common issues that have been fixed so far, and propose ways to prevent such errors from occurring in the future.

1. Scalability Issue: These issues are related to features not scaling well on production. For example, exploration state data may be too large for some explorations and prevent further changes from being made, or suggestion models being fetched for review may exceed the query limit, resulting in an incomplete result.

Preventive action:

- During design doc reviews, ask questions about performance and scalability when dealing with large amounts of data. Examples include: "How big do you expect X to get?", "Are there any limits to the size?", and "What happens when it reaches this limit and is it possible to handle it?"
- Implement a functionality to generate dummy data for testing locally.

2. Corner Cases: These are user-facing issues where certain corner case scenarios are not considered (e.g. empty lists, undefined objects). Examples include errors caused by deleting a hint that is active in the editor view. These issues can be tricky to debug due to lack of reproduction steps.

Preventive action:

- Use strongly typed frontend to help catch some of these issues.
- Encourage developers to test edge cases and use template users in testing to uncover and test edge case scenarios.

3. Supporting old browsers: Certain JS functionality may not be supported in older browsers. Examples include using String.prototype.match, which lacks support in IE browsers.

Preventive action:

- Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version/JS standard.

4. Logic errors: These are logical errors in the code. Examples include "this" scope referring to the global scope when it should refer to a local one. These issues are quite rare.

Preventive action:

- Flag these issues during code reviews, as they often happen when developers are refactoring existing code but haven't fully tested their changes.

5. Data-related Issues: These issues result in corrupt or missing data on the server. They can be hard to debug and involve complicated one-off jobs to be run. Examples include missing exploration stats models on the production server.

Preventive action:

- When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
- Test with foreign/special characters.
- Audit jobs need to account for all instances or occurrences of the relevant item on the server, sufficient written proof/explanation about the "thoroughness" of the audit can be added as a comment in the audit job code.

6. Race Conditions / Slow Network Issues: These errors happen when interacting too fast with the application or when the network is too slow. Examples include clicking on the "content" tab in the translation editor before it has fully loaded.

Preventive action:

- For code that has a lot of "async" functionality, encourage authors to try throttling their network and/or interacting with the UI before it has completely loaded. This can help catch potential race conditions.

7. Related to a migration: One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.

Preventive action:

- Documentation: Record common issues in the relevant migration page when they are identified.

8. Errors that didn't need to be logged: Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors.

Preventive action:
- Code / review: When adding logging statements or throwing errors that are not handled, always consider what scenarios and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the same (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)

In conclusion,  by implementing these preventive actions, it is possible to ensure a smoother experience for users and reduce the likelihood of server errors.

**Comment by Patel-Muhammad** [comment]
@kevintab95 can you please review my solution?

> ## @U8NWXD Sorry, my apologies.
> ### I will do the following:
> **1. Create a Server-errors-and-preventions.md file.**
> 
> **2. I will write about server errors from the [link](https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.4nsx9mmay1nn)**
> 
> **3. I will write the following in the .md file:**
> 
> This document presents an overview of different categories of server errors, based on errors that have been fixed in the past, and proposes ways to prevent such errors from occurring in the future. The categories include: Data-related issues, Scalability issues, Corner cases, Supporting old browsers, Logic errors, Race conditions/issues on slow networks, and Errors that didn't need to be logged. For each category, an example of an error is provided and preventive actions are suggested.
> 
> Server Errors are a common issue faced by developers and can be caused by a variety of factors. In this document, we will present the main categories of server errors based on common issues that have been fixed so far, and propose ways to prevent such errors from occurring in the future.
> 
> 1. Scalability Issue: These issues are related to features not scaling well on production. For example, exploration state data may be too large for some explorations and prevent further changes from being made, or suggestion models being fetched for review may exceed the query limit, resulting in an incomplete result.
> 
> Preventive action:
> 
> * During design doc reviews, ask questions about performance and scalability when dealing with large amounts of data. Examples include: "How big do you expect X to get?", "Are there any limits to the size?", and "What happens when it reaches this limit and is it possible to handle it?"
> * Implement a functionality to generate dummy data for testing locally.
> 
> 2. Corner Cases: These are user-facing issues where certain corner case scenarios are not considered (e.g. empty lists, undefined objects). Examples include errors caused by deleting a hint that is active in the editor view. These issues can be tricky to debug due to lack of reproduction steps.
> 
> Preventive action:
> 
> * Use strongly typed frontend to help catch some of these issues.
> * Encourage developers to test edge cases and use template users in testing to uncover and test edge case scenarios.
> 
> 3. Supporting old browsers: Certain JS functionality may not be supported in older browsers. Examples include using String.prototype.match, which lacks support in IE browsers.
> 
> Preventive action:
> 
> * Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version/JS standard.
> 
> 4. Logic errors: These are logical errors in the code. Examples include "this" scope referring to the global scope when it should refer to a local one. These issues are quite rare.
> 
> Preventive action:
> 
> * Flag these issues during code reviews, as they often happen when developers are refactoring existing code but haven't fully tested their changes.
> 
> 5. Data-related Issues: These issues result in corrupt or missing data on the server. They can be hard to debug and involve complicated one-off jobs to be run. Examples include missing exploration stats models on the production server.
> 
> Preventive action:
> 
> * When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
> * Test with foreign/special characters.
> * Audit jobs need to account for all instances or occurrences of the relevant item on the server, sufficient written proof/explanation about the "thoroughness" of the audit can be added as a comment in the audit job code.
> 
> 6. Race Conditions / Slow Network Issues: These errors happen when interacting too fast with the application or when the network is too slow. Examples include clicking on the "content" tab in the translation editor before it has fully loaded.
> 
> Preventive action:
> 
> * For code that has a lot of "async" functionality, encourage authors to try throttling their network and/or interacting with the UI before it has completely loaded. This can help catch potential race conditions.
> 
> 7. Related to a migration: One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.
> 
> Preventive action:
> 
> * Documentation: Record common issues in the relevant migration page when they are identified.
> 
> 8. Errors that didn't need to be logged: Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors.
> 
> Preventive action:
> 
> * Code / review: When adding logging statements or throwing errors that are not handled, always consider what scenarios and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the same (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)
> 

> In conclusion, by implementing these preventive actions, it is possible to ensure a smoother experience for users and reduce the likelihood of server errors.

@kevintab95 can you please review the wiki?

**Comment by vaishnavi192** [comment]
is this issue still open? can I work on it? @kevintab95 @U8NWXD

**Comment by vaishnavi192** [comment]
Hey @U8NWXD @kevintab95 
I created a server-error-and-solutions.md page.
Followed the oppia coding style and instructions to form a new page. If it seems right, kindly assign me this issue! 
![image](https://github.com/oppia/oppia-web-developer-docs/assets/117008958/b92094a3-342e-4bf7-82a2-79211d32f726)
![image](https://github.com/oppia/oppia-web-developer-docs/assets/117008958/15f2f338-c07b-4149-af9d-2ff1598bfb87)
![image](https://github.com/oppia/oppia-web-developer-docs/assets/117008958/31d9bdd0-50ef-43b2-b196-646883a2aeb6)

**Comment by seanlip** [comment]
@vaishnavi192 Seems like a good start, thanks. One note though, we should make sure this page is discoverable by developers. I suggest putting it in the sidebar, under "Coding Guidelines", just below the "Revert and Regression Policy" bullet point, Maybe we can call it "Coding safely to avoid server errors".

I'll assign you to this, feel free to create a PR. Thanks!

**Comment by vaishnavi192** [comment]
Hey @seanlip yes I a created a pr with required changes. Have a look at it. Thanks!

**Comment by seanlip** [comment]
This was fixed in https://github.com/oppia/oppia-web-developer-docs/pull/242. Closing.
